db/migrate/20190624084227_change_housing_stock_inputs_to_percentage.rb

```diff
@@ -0,0 +1,78 @@
+class ChangeHousingStockInputsToPercentage < ActiveRecord::Migration[5.1]
+  INPUTS = [
+    'apartments',
+    'corner_houses',
+    'detached_houses',
+    'semi_detached_houses',
+    'terraced_houses'
+  ].freeze
+
+  def change
+    reversible do |dir|
+      dir.up do
+        update_scenarios(INPUTS, 'number_of', 'share')
+      end
+
+      dir.down do
+        update_scenarios(
+          INPUTS.transform_values { |v| 1 / v },
+          'share',
+          'number_of'
+        )
+      end
+    end
+  end
+
+  private
+
+  def update_scenarios(collection, from_prefix, to_prefix)
+    total = scenarios.count
+    changed = 0
+
+    say "Checking and migrating #{total} scenarios"
+
+    scenarios.find_each.with_index do |scenario, index|
+      updated = collection.reduce(false) do |memo, (key)|
+        update_input(
+          scenario,
+          "households_#{from_prefix}_#{key}",
+          "households_#{key}_#{to_prefix}"
+        ) || memo
+      end
+
+      if updated
+        scenario.save(validate: false)
+        changed += 1
+      end
+
+      say "#{index + 1}/#{total} (#{changed} migrated)" if (index % 1000).zero?
+    end
+
+    say "#{total}/#{total} (#{changed} migrated)"
+  end
+
+  # Changes an input in a `scenario` from the `from` key to the `to` key,
+  # which is the percentage of the housing type compared to the total number of residences
+  #
+  # If the `from` key is not present, returns a falsey value.
+  def update_input(scenario, from, to)
+    number_of_residences = INPUTS.sum do |key|
```

Comment:
We could move the number_of_residences calculation up to `update_scenarios`, below: `scenarios.find_each.with_index` and pass in the value to `update_scenario` to avoid having to calculate it five times per scenario. Update input would need to take an extra parameter so that it could be given the value:

```ruby
def update_input(scenario, from, to, number_of_residences)
  # ...
end
```